File: src/sender_sup.erl
-module(sender_sup).
-export([start/0, init/1, start_worker/2]).
-behaviour(supervisor).
 
start() ->
    supervisor:start_link(?MODULE, []).
 
init([]) ->
    MaxRestart = 5,
    MaxTime = 3600,
    {ok, {{simple_one_for_one, MaxRestart, MaxTime},
	  [{sender,
	    {sender, start, []},
	    temporary, 5000, worker, [sender]}]}}.

start_worker(SenderPid, URL) ->
    supervisor:start_child(SenderPid, [URL]).
    

File: lib/sender_sup.ex
defmodule Drumbeat.SenderSup do
  use Supervisor

  # API
  def start_worker(sender_sup, uuid, request) do
    {:ok, _child} = Supervisor.start_child(sender_sup, [self(), uuid, request])
    :ok
  end

  def start_link do
    Supervisor.start_link(__MODULE__, :ok)
  end

  @worker_name Drumbeat.Sender

  def init(:ok) do
    children = [
      worker(@worker_name, [], restart: :transient)
    ]
    supervise(children,
              strategy: :simple_one_for_one,
              max_restarts: 5,
              max_seconds: 3600,
    )
  end
end
